src/tools/w3dviewer/w3dviewdoc.cpp

```diff
@@ -0,0 +1,957 @@
+    m_time(0.0f)
+{
+    m_animateCamera = theApp.GetProfileInt("Config", "AnimateCamera", 0) == 1;
+    m_resetCamera = theApp.GetProfileInt("Config", "ResetCamera", 1) == 1;
+}
+
+CW3DViewDoc::~CW3DViewDoc()
+{
+    Free();
+}
+
+BOOL CW3DViewDoc::OnNewDocument()
+{
+    BOOL result = CDocument::OnNewDocument();
+
+    if (result) {
+        m_loadedFiles.Delete_All();
+        m_forceCameraReset = true;
+
+        if (m_scene != nullptr && m_model != nullptr) {
+            RemoveRenderObject(m_model);
+        }
+
+        if (m_scene != nullptr) {
+            m_scene->Remove_All_LOD_Objects();
+            m_scene->Set_Fog_Color(m_fogColor);
+        }
+
+        if (m_model != nullptr) {
+            Ref_Ptr_Release(m_animation);
+            Ref_Ptr_Release(m_model);
+        }
+
+        CMainFrame *frame = reinterpret_cast<CMainFrame *>(AfxGetMainWnd());
+
+        if (frame != nullptr) {
+            CDataTreeView *view = reinterpret_cast<CDataTreeView *>(frame->m_splitter.GetPane(0, 0));
+
+            if (view != nullptr) {
+                view->GetTreeCtrl().DeleteItem(TVI_ROOT);
+                view->AddCategories();
+            }
+        }
+
+        Free();
+        W3DAssetManager::Get_Instance()->Free_Assets();
+
+        return TRUE;
+    }
+
+    return result;
+}
+
+BOOL CW3DViewDoc::OnOpenDocument(LPCTSTR lpszPathName)
+{
+    BOOL result = CDocument::OnOpenDocument(lpszPathName);
+    if (result) {
+        CGraphicView *view = GetCurrentGraphicView();
+
+        if (view != nullptr) {
+            view->EnableRendering(false);
+        }
+
+        LoadFile(lpszPathName);
+
+        CMainFrame *frame = reinterpret_cast<CMainFrame *>(AfxGetMainWnd());
+
+        if (frame != nullptr) {
+            CDataTreeView *tree = reinterpret_cast<CDataTreeView *>(frame->m_splitter.GetPane(0, 0));
+
+            if (tree != nullptr) {
+                tree->AddRenderObjects();
+            }
+        }
+
+        if (view != nullptr) {
+            view->EnableRendering(true);
+        }
+
+        return TRUE;
+    }
+
+    return result;
+}
+
+void CW3DViewDoc::Free()
+{
+    if (m_textureScene != nullptr) {
+        if (m_backgroundBitmap != nullptr) {
+            m_backgroundBitmap->Remove();
+        }
+
+        Ref_Ptr_Release(m_textureScene);
+    }
+
+    if (m_backgroundScene != nullptr) {
+        if (m_backgroundObject != nullptr) {
+            m_backgroundObject->Remove();
+        }
+
+        Ref_Ptr_Release(m_backgroundScene);
+    }
+
+    if (m_scene != nullptr) {
+        if (m_model != nullptr) {
+            RemoveRenderObject(m_model);
+        }
+
+        if (m_light != nullptr) {
+            RemoveRenderObject(m_light);
+        }
+
+        m_scene->Remove_All_LOD_Objects();
+        Ref_Ptr_Release(m_scene);
+    }
+
+    Ref_Ptr_Release(m_textureCamera);
+    Ref_Ptr_Release(m_docCamera);
+    Ref_Ptr_Release(m_backgroundBitmap);
+    Ref_Ptr_Release(m_light);
+
+    if (m_model != nullptr) {

```

Comment:
With this test in place it looks like `m_animation` could leak. Maybe omit this test, because `Ref_Ptr_Release` already tests null?

Reply:
m_animation would only be set if m_model is also set.